extension/helper/cart.js

```diff
@@ -82,38 +82,49 @@ const handleCartError = async (err, checkoutCartItems, cartId, context) => {
  * @param {Object} lineItems
  * @returns {Array}
  */
-function getOutOfStockLineItemIds (lineItems) {
-  const itemsToDelete = []
+function getLineItemIdsWithQuantityErrors (lineItems) {
+  const itemsToUpdate = []
 
   for (let itemId in lineItems) {
     Object.entries(lineItems[itemId]).forEach(([errorType, errors]) => {
-      if (errors.find(error => error.code === 'not_enough_in_stock' &&
+      const quantityError = errors.find(error => error.code === 'not_enough_in_stock' &&
         error.options &&
-        error.options.remaining === 0
-      )) {
-        itemsToDelete.push(parseInt(itemId))
+        error.options.remaining >= 0
+      )
+      if (quantityError) {
+        itemsToUpdate.push({ id: parseInt(itemId), availableQuantity: quantityError.options.remaining })
       }
     })
   }
 
-  return itemsToDelete
+  return itemsToUpdate
 }
 
 /**
  * @param {Array} checkoutCartItems
- * @param {Array} itemsToDelete
+ * @param {Array} itemsToUpdate
  * @param {Error} error
  * @param {Object} error.errors
  * @param {Object} error.errors.line_items
  * @param {number} cartId
  * @param {SDKContext} context
  */
-async function fixCheckoutQuantities (checkoutCartItems, itemsToDelete, error, cartId, context) {
+async function fixCheckoutQuantities (checkoutCartItems, itemsToUpdate, error, cartId, context) {
   const shopifyApiRequest = new ShopifyApiRequest(context.config, context.log)
 
-  for (let itemId of itemsToDelete) {
-    checkoutCartItems.splice(itemId, 1)
-    delete error.errors.line_items[itemId]
+  // fix quantities first, to not mess with index
+  for (let item of itemsToUpdate) {
+    if (item.availableQuantity !== 0) {
```

Comment:
can it be just `>` or are we expecting some negative numbers in quantity? can you test if negative number can be set in shopify? if so, can u also create a node test for it as ur logic does not cover such cases.